builder/builder.go

```diff
@@ -0,0 +1,569 @@
+package builder
+
+import (
+	"fmt"
+	"strings"
+)
+
+type Builder struct {
+	spacing string // optional spacing [s]
+	newline string
+	err     error
+
+	quoted            pattern
+	locals            []local
+	inputs, selectors []expression
+	variants          []variant
+	patterns          []any
+}
+
+func New() *Builder {
+	return &Builder{
+		newline: "\n",
+		spacing: " ",
+	}
+}
+
+func (b *Builder) build() (string, error) {
+	if b.err != nil {
+		return "", b.err
+	}
+
+	if err := b.validate(); err != nil {
+		return "", err
+	}
+
+	var s string
+
+	for _, v := range b.inputs {
+		s += ".input" + b.spacing + v.String() + b.newline
+	}
+
+	for _, v := range b.locals {
+		s += ".local" + b.spacing + string(v.variable) + b.spacing + "=" + b.spacing + v.expr.String() + b.newline
+	}
+
+	for _, v := range b.patterns {
+		switch v := v.(type) {
+		case string:
+			s += v
+		case *expression:
+			s += v.String()
+		default:
+			return "", fmt.Errorf("unsupported pattern type: %T", v)
+		}
+	}
+
+	for i, v := range b.selectors {
+		switch i {
+		case 0:
+			s += ".match" + b.spacing + v.String() + " "
+		case len(b.selectors) - 1: // newline after the last selector
+			s += v.String() + b.newline
+		default:
+			s += v.String() + " "
+		}
+	}
+
+	for i, v := range b.variants {
+		s += v.String()
+
+		if i != len(b.variants)-1 {
+			s += b.newline
+		}
+	}
+
+	if len(b.quoted) > 0 {
+		s += b.quoted.String()
+	}
+
+	return s, nil
+}
+
+func (b *Builder) MustBuild() string {
+	s, err := b.build()
+	if err != nil {
+		panic(err)
+	}
+
+	return s
+}
+
+func (b *Builder) validate() error {
+	if len(b.variants) == 0 && b.quoted == nil &&
+		(len(b.inputs) > 0 || len(b.locals) > 0 || len(b.selectors) > 0) {
+		return fmt.Errorf("complex message MUST include complex body")
+	}
+
+	if len(b.variants) > 0 {
+		if b.quoted != nil {
+			return fmt.Errorf("complex message MUST have single complex body")
+		}
+
+		if len(b.selectors) == 0 {
+			return fmt.Errorf("matcher MUST have at least one selector")
+		}
+
+		if !hasCatchAllVariant(b.variants) {
+			return fmt.Errorf("matcher MUST have at least one variant with all catch-all keys")
+		}
+	}
+
+	if len(b.selectors) > 0 && len(b.variants) == 0 {
+		return fmt.Errorf("matcher MUST have at least one variant")
+	}
+
+	return nil
+}
+
+func (b *Builder) Newline(s string) *Builder {
+	if b.err != nil {
+		return b
+	}
+
+	b.newline = s
+
+	return b
+}
+
+func (b *Builder) Spacing(s string) *Builder {
+	if b.err != nil {
+		return b
+	}
+
+	b.spacing = s
+
+	return b
+}
+
+func (b *Builder) Text(s string) *Builder {
+	if b.err != nil {
+		return b
+	}
+
+	if b.IsEmpty() {
+		if hasSimpleStart(s) || s == "" {
+			b.patterns = append(b.patterns, textEscape(s))
```

Comment:
Please serialize values in the `Build()`. Do not do it early. Makes it harder to read and maintain the code.

`Build()` - concatenates and serializes values, and builds final value.

Reply:
I mean, use raw values in build steps.